### PR TITLE
Implemented Web3 integration logic

### LINKED ISSUE #12: Web3 Integration
#### User Story

As a user, I want to be able to connect my web3 wallet to the website. So that I can buy the portfolio tokens from a Uniswap market in Kovan.

#### Type

- [x] Enhancement
- [ ] Maintenance
- [ ] Refactor

#### Description
To be able to buy tokens on Uniswap in Kovan there should be a web3 provider in the context.

#### Solution

- [x] Add the web3-react context
- [x] Pop up the provider selection modal on clicking the connect button
- [x] Create a provider selection modal
- [x] Connect the web3 provider upon selection of the web3 provider
- [x] Connect Kovan Metamask when the user clicks connect

#### Definition of Done

- [x] A user is able to connect their web3 wallet.
- [x] The user can connect Kovan Metamask
- [ ] The user gets feedback that Kovan Metamask is connected
- [x] The user is notified if he is in the wrong network

### PR DESCRIPTION
**Pull request by MickdeGraaf**
##### Description
Solved #12 

##### Checklist
- [x] Changes don't break existing behavior
- [x] Tests coverage hasn't decreased

##### Testing
Adds new functionality but leaves existing code the same.

##### Refers/Fixes
Fixes : #12

### CODE REVIEW

**Comment by MickdeGraaf** [comment]
Modal does some weird flashes on re-rendering. I will discuss with @dimarconicola if he knows fixes to this issue or use some other modal system.

This PR is only for the web3 connecting logic, ui does not represent how it is supposed to be.

File: src/Components/WalletModal/index.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            setPendingError(true);
+            setErrorMessage(e.message);
+        })
+    }
+
+    // close modal when a connection is successful
+    const activePrevious = usePrevious(active)
+    const connectorPrevious = usePrevious(connector)
+    useEffect(() => {
+        if (walletModalOpen && ((active && !activePrevious) || (connector && connector !== connectorPrevious && !error))) {
```
Comment (Alexintosh) [inline src/Components/WalletModal/index.js:82]: maybe let's format this a bit better

File: src/Components/WalletModal/index.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            isActive={walletModalOpen} // required
+            closeModal={toggleWalletModal} // required
+            showAnimation={true}
+            modalBoxStyle={{
+                width: "90%",
+                maxWidth: 600,
+                padding: "5%"
+            }}
+            >
+            <IF what={walletView == WALLET_VIEWS.OPTIONS}>
```
Comment (Alexintosh) [inline src/Components/WalletModal/index.js]: === please

File: src/Components/WalletModal/index.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                <WalletOption onClick={() => { tryActivation(SUPPORTED_WALLETS.METAMASK.connector) }} wallet={SUPPORTED_WALLETS.METAMASK} />
+            </IF>
+            <IF what={walletView == WALLET_VIEWS.PENDING}>
```
Comment (Alexintosh) [inline src/Components/WalletModal/index.js:103]: Same ===

File: src/Components/WalletModal/index.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+                {pendingError && 
+                    <>
+                        <div>ERROR CONNECTING</div>
+                        {errorMessage}
+                        <PrimaryButton onClick={() => {setWalletView(WALLET_VIEWS.OPTIONS)}}>GO BACK</PrimaryButton>
+                    </>
+                }
+            </IF>
+            <IF what={walletView == WALLET_VIEWS.ACCOUNT}>
```
Comment (Alexintosh) [inline src/Components/WalletModal/index.js:120]: ===

File: src/contexts/Application.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      [WALLET_MODAL_OPEN]: false
+    })
+  
+    const toggleWalletConnectionModal = useCallback(() => {
+      dispatch({ type: TOGGLE_WALLET_CONNECTION_MODAL })
+    }, [])
+  
+    return (
+      <ApplicationContext.Provider
+        value={useMemo(() => [state, { toggleWalletConnectionModal }], [
```
Comment (Alexintosh) [inline src/contexts/Application.js:36]: Does `useMemo` memoize data? Just courious

**Review by Alexintosh — APPROVED** [review]

File: src/Components/WalletModal/index.js (hunk 1 above)
Comment (MickdeGraaf) [reply src/Components/WalletModal/index.js:82]: https://github.com/Alexintosh/PIE-landing/pull/30/commits/7fb47dbb2650ae4701e804eaef777dfd750e65f9 This should make it a bit more readable

File: src/contexts/Application.js (hunk 5 above)
Comment (MickdeGraaf) [reply src/contexts/Application.js:36]: Yes it does. In this case I would think it memoizes the default context value I think

**Review by dmvt — APPROVED** [review]

**Pull request merged**